nixos/roles/kubernetes/master.nix

```diff
@@ -0,0 +1,292 @@
+        "addonmanager.kubernetes.io/mode" = "Reconcile";
+      };
+      name = "kubernetes-dashboard";
+      namespace  = "kube-system";
+    };
+    spec = {
+      clusterIP = cfg.dashboardClusterIP;
+      ports = [{
+        port = 443;
+        targetPort = 8443;
+      }];
+      selector.k8s-app = "kubernetes-dashboard";
+    };
+  };
+
+in
+{
+  options = {
+    flyingcircus.roles.kubernetes-master = {
+
+      enable = lib.mkEnableOption "Enable Kubernetes Master (only one per RG; experimental)";
+
+      dashboardClusterIP = lib.mkOption {
+        default = "10.0.0.250";
+      };
+
+    };
+  };
+
+  config = lib.mkIf cfg.enable {
+
+    # Create kubeconfigs for all users with an admin cert (sudo-srv).
+    environment.etc =
+      lib.mapAttrs'
+        (n: v: lib.nameValuePair
+          "/kubernetes/${n}.kubeconfig"
+          { source = mkUserKubeConfig v; })
+        allCerts;
+
+    environment.shellInit = lib.mkAfter ''
+      # Root uses the cluster-admin cert generated by NixOS.
+      if [[ $UID == 0 ]]; then
+        export KUBECONFIG=/etc/kubernetes/cluster-admin.kubeconfig
+      else
+        export KUBECONFIG=/etc/kubernetes/$USER.kubeconfig
+      fi
+    '';
+
+    environment.systemPackages = with pkgs; [
+      kubectl
+      kubernetesEtcdctl
+      kubernetesMakeKubeconfig
+      sensu-plugins-kubernetes
+    ];
+
+    # Policy routing interferes with virtual ClusterIPs handled by kube-proxy, disable it.
+    flyingcircus.network.policyRouting.enable = false;
+
+    flyingcircus.activationScripts.kubernetes-apitoken =
+      lib.stringAfter [ "users" ] ''
+        mkdir -p /var/lib/cfssl
+        umask 077
+        tok=/var/lib/cfssl/apitoken.secret
+        echo ${master.password} | md5sum | head -c32 > $tok
+        chown cfssl $tok && chmod 400 $tok
+      '';
+
+    flyingcircus.services.sensu-client.checks = let
+      bin = "${pkgs.sensu-plugins-kubernetes}/bin";
+      cfg = "--kube-config /etc/kubernetes/sensu.kubeconfig";
+    in
+    {
+      kube-apiserver = {
+        notification = "Kubernetes API server is not working";
+        command = ''
+          ${bin}/check-kube-apiserver-available.rb ${cfg}
+        '';
+      };
+
+      kube-dashboard = {
+        notification = "Kubernetes dashboard is not working";
+        command = ''
+          ${bin}/check-kube-service-available.rb -l kubernetes-dashboard ${cfg}
+        '';
+      };
+
+      kube-dns = {
+        notification = "Kubernetes DNS is not working";
+        command = ''
+          ${bin}/check-kube-service-available.rb -l kube-dns ${cfg}
+        '';
+      };
+
+    };
+
+    networking.firewall = {
+      allowedTCPPorts = [ apiserverPort ];
+    };
+
+    services.kubernetes = {
+      addons.dashboard.enable = true;
+      addonManager.addons.kubernetes-dashboard-svc = lib.mkForce dashboardSvc;
+      apiserver.extraSANs = apiserverAddresses;
+      # Changing the masterAddress is tricky and requires manual intervention.
```

Comment:
master is a precious thing and can't be moved around easily. please note this in the docs